File: src/de/umr/fixcon/graphFunctions/standardFunctions/MaxDegreeFunction.java
package de.umr.fixcon.graphFunctions.standardFunctions;

import com.google.common.graph.Graph;
import de.umr.fixcon.graphFunctions.AbstractGraphFunction;
import de.umr.fixcon.graphFunctions.GraphFunction;

import java.util.OptionalInt;

import static com.google.common.base.Preconditions.checkArgument;

public class MaxDegreeFunction extends AbstractGraphFunction implements GraphFunction {

    @Override
    public boolean isEdgeMonotone() {
        return true;
    }

    @Override
    public double apply(final Graph<Integer> g, final int... parameters) {
        OptionalInt result = degreeStream(g).max();
        checkArgument(result.isPresent(), "Graph does not contain any edges");
        return result.getAsInt();
    }

    @Override
    public double optimum(int size) {
        return size - 1;
    }
}

File: src/de/umr/fixcon/graphFunctions/standardFunctions/MaxDegreeFunction.kt
package de.umr.fixcon.graphFunctions.standardFunctions

import de.umr.core.dataStructures.VertexOrderedGraph
import de.umr.core.degreeSequence
import de.umr.core.vertexCount
import de.umr.fixcon.graphFunctions.GraphFunction
import org.jgrapht.Graph
import org.jgrapht.graph.DefaultEdge

/**Returns the maximum degree of all vertices in this graph.*/
object MaxDegreeFunction :  GraphFunction {
    override val isEdgeMonotone: Boolean = true

    override fun completeAdditionBound(subgraph: VertexOrderedGraph<Int>, targetSize: Int, args: List<Int>) =
            targetSize- subgraph.vertexCount

    override fun eval(g: Graph<Int, DefaultEdge>, args: List<Int>) = g.degreeSequence.max()!!

    override fun globalUpperBound(graphSize: Int)  : Int {
        require(graphSize >= 0)
        return graphSize-1
    }
}